### PR TITLE
update operator config observation status

### PR DESCRIPTION
**Pull request by sanchezl**
enabled config observation status
switched to using Listers

### CODE REVIEW

File: pkg/operator/observe_config.go [hunk 1]
```diff
@@ -90,69 +89,111 @@ func (c ConfigObserver) sync() error {
 	// don't worry about errors
 	currentConfig := map[string]interface{}{}
 	json.NewDecoder(bytes.NewBuffer(operatorConfig.Spec.ObservedConfig.Raw)).Decode(&currentConfig)
-	if reflect.DeepEqual(currentConfig, observedConfig) {
-		return nil
+
+	var errs []error
```
Comment (mfojtik) [inline pkg/operator/observe_config.go]:
```suggestion
	var (
	  errs []error
	  observedConfigs []map[string]interface{}
	)
```

File: pkg/operator/observe_config.go [hunk 2]
```diff
@@ -90,69 +89,111 @@ func (c ConfigObserver) sync() error {
...
+	var observedConfigs []map[string]interface{}
+	for _, i := range rand.Perm(len(c.observers)) {
+		var currErrs []error
```
Comment (mfojtik) [inline pkg/operator/observe_config.go]:
```suggestion
```

Comment (mfojtik) [reply pkg/operator/observe_config.go]: No need to declare this when you re-declaring below?

File: pkg/operator/observe_config.go [hunk 3]
```diff
@@ -90,69 +89,111 @@ func (c ConfigObserver) sync() error {
...
-		return err
+	mergedObservedConfig := map[string]interface{}{}
+	for _, observedConfig := range observedConfigs {
+		mergo.Merge(&mergedObservedConfig, observedConfig)
+	}
+
+	if !equality.Semantic.DeepEqual(currentConfig, mergedObservedConfig) {
+		glog.Infof("writing updated observedConfig: %v", diff.ObjectDiff(operatorConfig.Spec.ObservedConfig.Object, mergedObservedConfig))
+		operatorConfig.Spec.ObservedConfig = runtime.RawExtension{Object: &unstructured.Unstructured{Object: mergedObservedConfig}}
+		operatorConfig, err = c.operatorConfigClient.KubeControllerManagerOperatorConfigs().Update(operatorConfig)
```
Comment (mfojtik) [inline pkg/operator/observe_config.go]:
```suggestion
		updatedOperatorConfig, err := c.operatorConfigClient.KubeControllerManagerOperatorConfigs().Update(operatorConfig)
		if err != nil {
 			errs = append(errs, fmt.Errorf("kubecontrollermanageroperatorconfigs/instance: error writing updated observed config: %v", err))
 		} else {
 		    operatorConfig = updatedOperatorConfig
 		}
```

Comment (mfojtik) [reply pkg/operator/observe_config.go]: interesting... suggestions only works for a single line changes :-)

Comment (sanchezl) [reply pkg/operator/observe_config.go]: Thanks. In an attempt to explain myself here, I took a closer look at the `Update` api source and realized it wasn't doing what I expected. I've re-worked the logic and added a unit test for the error condition.

File: pkg/operator/observe_config.go (hunk 2 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: :+1:

File: pkg/operator/observe_config.go (hunk 1 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: :+1:

File: pkg/operator/observe_config.go [hunk 4]
```diff
@@ -4,37 +4,51 @@ import (
...
+// while attempting to generate the observedConfig should be returned in the errs slice.
+type observeConfigFunc func(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error)
 
 type ConfigObserver struct {
 	operatorConfigClient operatorconfigclientv1alpha1.KubecontrollermanagerV1alpha1Interface
 
-	kubeClient   kubernetes.Interface
-	clientConfig *rest.Config
+	listers         Listers
+	configmapSynced cache.InformerSynced
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: use the slice of sync funcs like I did in kube-apiserver-operator

File: pkg/operator/observe_config.go [hunk 5]
```diff
@@ -90,69 +89,114 @@ func (c ConfigObserver) sync() error {
...
+		}
 	}
 
 	return nil
 }
 
-// observeClusterConfig observes cloud provider configuration from
-// cluster-config-v1 in order to configure kube-controller-manager's cloud
+// observeCloudProviderNames observes cloud provider configuration from
+// cluster-config-corelistersv1 in order to configure kube-controller-manager's cloud
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: bad sed

File: pkg/operator/observe_config.go [hunk 6]
```diff
@@ -90,69 +89,114 @@ func (c ConfigObserver) sync() error {
...
 // provider.
-func observeClusterConfig(kubeClient kubernetes.Interface, clientConfig *rest.Config, observedConfig map[string]interface{}) (map[string]interface{}, error) {
-	clusterConfig, err := kubeClient.CoreV1().ConfigMaps("kube-system").Get("cluster-config-v1", metav1.GetOptions{})
+func observeCloudProviderNames(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error) {
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: no named return values.

File: pkg/operator/observe_config.go [hunk 7]
```diff
@@ -4,37 +4,51 @@ import (
...
+
+type Listers struct {
+	configmapLister corelistersv1.ConfigMapLister
+}
+
+// observeConfigFunc observes configuration and returns the observedConfig. This function should not return an
+// observedConfig that would cause the service being managed by the operator to crash. For example, if a required
+// configuration key cannot be observed, consider reusing the configuration key's previous value. Errors that occur
+// while attempting to generate the observedConfig should be returned in the errs slice.
+type observeConfigFunc func(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error)
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: `expectedConfig`

File: pkg/operator/observe_config.go [hunk 8]
```diff
@@ -90,69 +89,114 @@ func (c ConfigObserver) sync() error {
...
 	if errors.IsNotFound(err) {
-		glog.Warningf("cluster-config-v1 not found in the kube-system namespace")
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: not found")
+		return
 	}
 	if err != nil {
-		return observedConfig, err
+		errs = append(errs, err)
+		if len(currentCloudProvider) > 0 {
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: I think the construction in https://github.com/openshift/cluster-kube-apiserver-operator/pull/116 is easier for "use old value"

File: pkg/operator/observe_config.go [hunk 9]
```diff
@@ -90,69 +89,114 @@ func (c ConfigObserver) sync() error {
...
 	//       aws: {}
-	platform, ok := installConfig["platform"].(map[string]interface{})
-	if !ok {
-		glog.Warningf("Unable to parse install-config: %s", err)
-		return observedConfig, nil
-	}
-
-	cloudProvider := ""
-	switch {
-	case platform["aws"] != nil:
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: looks like you've lost this and made it much harder to extend to our other cloud providers.  please restore.

File: pkg/operator/observe_config.go [hunk 10]
```diff
@@ -90,69 +89,114 @@ func (c ConfigObserver) sync() error {
 	// don't worry about errors
 	currentConfig := map[string]interface{}{}
 	json.NewDecoder(bytes.NewBuffer(operatorConfig.Spec.ObservedConfig.Raw)).Decode(&currentConfig)
-	if reflect.DeepEqual(currentConfig, observedConfig) {
-		return nil
+
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: you've moved this block to a few spots now.  Got a generic construct for us to use?

File: pkg/operator/observe_config.go (hunk 5 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: :+1:

File: pkg/operator/observe_config.go (hunk 4 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: :+1:

File: pkg/operator/observe_config.go (hunk 6 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: :+1:

File: pkg/operator/observe_config.go (hunk 7 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: I see you renamed to `existingConfig` in **cluster-kube-api-server-operator**. Is that what you meant?

Comment (deads2k) [reply pkg/operator/observe_config.go]: > I see you renamed to `existingConfig` in **cluster-kube-api-server-operator**. Is that what you meant?

yeah

Comment (deads2k) [reply pkg/operator/observe_config.go]: Comments here still outstanding

File: pkg/operator/observe_config.go [hunk 11]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: install-config not found")
+		return observedConfig, errs
 	}
 	installConfig := map[string]interface{}{}
 	err = yaml.Unmarshal([]byte(installConfigYaml), &installConfig)
 	if err != nil {
 		glog.Warningf("Unable to parse install-config: %s", err)
-		return observedConfig, nil
+		return observedConfig, errs
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: use the previous value

File: pkg/operator/observe_config.go [hunk 12]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
 	platform, ok := installConfig["platform"].(map[string]interface{})
-	if !ok {
-		glog.Warningf("Unable to parse install-config: %s", err)
-		return observedConfig, nil
-	}
-
-	cloudProvider := ""
 	switch {
+	case !ok:
+		glog.Warning("configmap/cluster-config-v1.kube-system: install-config/platform not found")
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: wrong message

File: pkg/operator/observe_config.go [hunk 13]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
+		return observedConfig, errs
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: return previous value

File: pkg/operator/observe_config.go [hunk 14]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
 	}
 
 	// extract needed values
 	//  data:
 	//   install-config:
 	//     platform:
 	//       aws: {}
+	// only aws supported for now
+	var cloudProvider string
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: `cloudProvider := ""`

File: pkg/operator/observe_config.go [hunk 15]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
+	case !ok:
+		glog.Warning("configmap/cluster-config-v1.kube-system: install-config/platform not found")
+		return observedConfig, errs
 	case platform["aws"] != nil:
 		cloudProvider = "aws"
 	default:
-		glog.Warningf("No recognized cloud provider found in install-config/platform")
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: no recognized cloud provider platform found")
+		return observedConfig, errs
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: previous value

File: pkg/operator/observe_config.go [hunk 16]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
 	switch {
+	case !ok:
+		glog.Warning("configmap/cluster-config-v1.kube-system: install-config/platform not found")
+		return observedConfig, errs
 	case platform["aws"] != nil:
 		cloudProvider = "aws"
 	default:
-		glog.Warningf("No recognized cloud provider found in install-config/platform")
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: no recognized cloud provider platform found")
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: this should be an error

File: pkg/operator/observe_config.go [hunk 17]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
 	installConfigYaml, ok := clusterConfig.Data["install-config"]
 	if !ok {
-		return observedConfig, nil
+		glog.Warning("configmap/cluster-config-v1.kube-system: install-config not found")
+		return observedConfig, errs
 	}
 	installConfig := map[string]interface{}{}
 	err = yaml.Unmarshal([]byte(installConfigYaml), &installConfig)
 	if err != nil {
 		glog.Warningf("Unable to parse install-config: %s", err)
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: you should add this to errors

File: pkg/operator/observe_config.go [hunk 18]
```diff
@@ -90,71 +97,126 @@ func (c ConfigObserver) sync() error {
...
 	return nil
 }
 
-// observeClusterConfig observes cloud provider configuration from
+// observeCloudProviderNames observes cloud provider configuration from
 // cluster-config-v1 in order to configure kube-controller-manager's cloud
 // provider.
-func observeClusterConfig(kubeClient kubernetes.Interface, clientConfig *rest.Config, observedConfig map[string]interface{}) (map[string]interface{}, error) {
-	clusterConfig, err := kubeClient.CoreV1().ConfigMaps("kube-system").Get("cluster-config-v1", metav1.GetOptions{})
+func observeCloudProviderNames(listers Listers, currentConfig map[string]interface{}) (map[string]interface{}, []error) {
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: existingConfig

File: pkg/operator/observe_config.go [hunk 19]
```diff
@@ -4,63 +4,79 @@ import (
...
+
+type Listers struct {
+	configmapLister corelistersv1.ConfigMapLister
+}
+
+// observeConfigFunc observes configuration and returns the observedConfig. This function should not return an
+// observedConfig that would cause the service being managed by the operator to crash. For example, if a required
+// configuration key cannot be observed, consider reusing the configuration key's previous value. Errors that occur
+// while attempting to generate the observedConfig should be returned in the errs slice.
+type observeConfigFunc func(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error)
```
Comment (deads2k) [inline pkg/operator/observe_config.go]: existingConfig

**Pull request merged**